### PR TITLE
add datetime2 support

### PR DESCRIPTION
**Pull request by avi-berkowitz-zocdoc**
@chris-smith-zocdoc @mustafa-rizvi-zocdoc @Carson-Carlisle-ZocDoc 

Our `AddScalarParameter` function uses `AddWithValue` under the hood, which infers `DbType`.  Because C# maps SQL's `datetime` and `datetime2` to C#'s `DateTime`, the default conversion of `DateTime` to a SQL type yields a `datetime`.  The results are twofold:

1) An out of range `DateTime` (eg, before year 1973) will throw an exception when it converts to `datetime`
2) A `DateTime` with data more precise than the max precision of `datetime` (I believe .003 second) will be truncated before being stored

Both of these issues hold true EVEN if the SQL procedure or table expects a `datetime2`.

The solution is to detect the type of value and if it's a `DateTime`, manually create a parameter, and set the `SqlDbType` to `SqlDbType.DateTime2`.  Even if the underlying SQL procedure expects a `datetime`, this will still convert to `datetime2`.  The results are the same as the two listed above.

In summary, this change shouldn't cause any new breakages if the underlying data type is `datetime`, but if the underlying data type is `datetime2`, this should prevent out of range exceptions and truncation.

### CODE REVIEW

**Comment by kyle-fritz-zocdoc** [comment]
🥇

File: ZocUtil.Database.Tests/SqlCommandHelperTests.cs [hunk 1]
```diff
@@ -28,6 +27,191 @@ public void AddScalarParameterUsesDbNullForNullValues()
...
+            {
+                var systemUnderTest = new SqlCommandHelper(query, conn);
+                systemUnderTest.AddScalarParameter(parameterName, dateTime);
+
+                // Act
+                var result = systemUnderTest.ExecuteScalar<DateTime>();
+
+                // Assert
+                Assert.IsNotNull(result);
+                Assert.AreEqual(dateTime, result);
```
Comment (mustafa-rizvi-zocdoc) [inline ZocUtil.Database.Tests/SqlCommandHelperTests.cs]: its unclear to me how you're verifying precision. Can you add a more explicit assert?

File: ZocUtil.Database.Tests/SqlCommandHelperTests.cs [hunk 2]
```diff
@@ -28,6 +27,191 @@ public void AddScalarParameterUsesDbNullForNullValues()
...
```
Comment (mustafa-rizvi-zocdoc) [inline ZocUtil.Database.Tests/SqlCommandHelperTests.cs]: nit: since i can't be sure that `DateTime.Parse` did not parse year 1100 as some non-null default, can we check year or something

**Comment by mustafa-rizvi-zocdoc** [comment]
magical 🌟 
👍

File: ZocUtil.Database.Tests/SqlCommandHelperTests.cs (hunk 2 above)
Comment (mustafa-rizvi-zocdoc) [inline ZocUtil.Database.Tests/SqlCommandHelperTests.cs]: nit: since i can't be sure that DateTime.Parse did not parse year 1100 as some non-null default, can we check year or something

Comment (avi-berkowitz-zocdoc) [inline ZocUtil.Database.Tests/SqlCommandHelperTests.cs]: @mustafa-rizvi-zocdoc added an assert

File: ZocUtil.Database.Tests/SqlCommandHelperTests.cs (hunk 1 above)
Comment (avi-berkowitz-zocdoc) [inline ZocUtil.Database.Tests/SqlCommandHelperTests.cs]: @mustafa-rizvi-zocdoc added assert to compare string to original string, which has the precision

**Comment by mustafa-rizvi-zocdoc** [comment]
:shipit:
